proposals/infra/INF-0004-validator-hashing.md

```diff
@@ -195,6 +163,93 @@ multiple-year-old preview compiler for generating final compiled shaders for a
 title. The high rate of bugs in preview compilers contributes to making this
 even less likely.
 
+## Detailed Design
+
+### D3D Runtime Behavior
+
+Starting with AgilitySDK version TBD, the runtime adopts new behavior for how 
+and when the shader hash is validated, including support for bytecode validation 
+in the debug layer.
+
+For shaders with the `PREVIEW_BYPASS` hash, the shader is allowed to execute 
+only if developer mode and the experimental feature 
+`D3D12ExperimentalShaderModels` is enabled, otherwise the runtime will produce
+an error.
+
+For shaders with the `BYPASS` hash, shaders can run without validating the hash 
+regardless of whether the machine is in developer mode as long as the shader 
+targets a supported non-experimental shader model.
+
+All other values of the hash are assumed to be real hash values.
+
+For shaders that contain a real hash value, the behavior is unchanged. The hash
+is verified and the shader executes as it does with existing versions of the 
+runtime.
+
+> Note: an important behavior change here is that shaders with zero'd hash data
+> will always be treated as invalid by the runtime and rejected.
+> 
+> An exception is for a period of time already existing preview shaders that used 
+> zero'd hash data will be treated like having the `PREVIEW_BYPASS` hash in that
+> preview runtimes will allow them to be used, at least until those preview shaders
+> and runtimes are no longer supported.  In contrast to `PREVIEW_BYPASS` shaders,
+> these legacy zero'd has shaders will not run have the validator invoked on them
+> by the runtime/debug layer since that wasn't expected of them.
+
+#### D3D Debug Layer Validation Control
+
+Apps can configure how the debug layer runtime validates bytecode passed to it. 
+
+The validation is the same code the compiler uses when it endorses bytecode at 
+compile time by applying a hash.  A copy of this validation code is in 
+the debug layer for it to use.  So it would typically be 
+redundant to validate again, but it can be useful to validate shaders that have 
+the `BYPASS` or `PREVIEW_BYPASS` hash, or have the option to force validation. 
+
+If the debug layer finds errors when running the bytecode validator it only 
+reports messages, but doesn't fail shader creation.
+
+The validation mode can be configured via ID3D12DebugDevice1::SetDebugParameter()`:
+
+```C++
+typedef enum D3D12_DEBUG_DEVICE_PARAMETER_TYPE
+{
+    ...
+    D3D12_DEBUG_DEVICE_PARAMETER_BYTECODE_VALIDATION_MODE   
+} D3D12_DEBUG_DEVICE_PARAMETER_TYPE;
+
+typedef enum D3D12_BYTECODE_VALIDATE_MODE
```

Comment:
For other values:

* `D3D12_DEBUG_DEVICE_PARAMETER_FEATURE_FLAGS` --> `D3D12_DEBUG_FEATURE`
* `D3D12_DEBUG_DEVICE_PARAMETER_GPU_BASED_VALIDATION_SETTINGS` --> `D3D12_DEBUG_DEVICE_GPU_BASED_VALIDATION_SETTINGS`
* `D3D12_DEBUG_DEVICE_PARAMETER_GPU_SLOWDOWN_PERFORMANCE_FACTOR` --> `D3D12_DEBUG_DEVICE_GPU_SLOWDOWN_PERFORMANCE_FACTOR`

Following this pattern, I think the enum should be called:

`D3D12_DEBUG_DEVICE_BYTECODE_VALIDATE_MODE` or `D3D12_DEBUG_BYTECODE_VALIDATE_MODE`

Reply:
Fixed